compiler-core/src/package_interface.rs

```diff
@@ -0,0 +1,596 @@
+    ///   4
+    ///   // This is a default implementation.
+    /// }
+    /// ```
+    /// The implementations for this function will look like this:
+    /// ```json
+    /// {
+    ///   gleam: true,
+    ///   uses_erlang_externals: true,
+    ///   uses_javascript_externals: false,
+    /// }
+    /// ```
+    /// - `gleam: true` means that the function has a pure Gleam implementation
+    ///   and thus it can be used on all Gleam targets with no problems.
+    /// - `uses_erlang_externals: true` means that the function will use Erlang
+    ///   external code when compiled to the Erlang target.
+    /// - `uses_javascript_externals: false` means that the function won't use
+    ///   JavaScript external code when compiled to JavaScript. The function can
+    ///   still be used on the JavaScript target since it has a pure Gleam
+    ///   implementation.
+    ///
+    /// Now have a look at this function:
+    ///
+    /// ```gleam
+    /// @external(javascript, "foo", "bar")
+    /// pub fn javascript_only() -> Int
+    /// ```
+    /// It's implementations field will look like this:
+    /// ```json
+    /// {
+    ///   gleam: false,
+    ///   uses_erlang_externals: false,
+    ///   uses_javascript_externals: true,
+    /// }
+    /// ```
+    /// - `gleam: false` means that the function doesn't have a pure Gleam
+    ///   implementations. This means that the function is only defined using
+    ///   externals and can only be used on some targets.
+    /// - `uses_erlang_externals: false` the function is not using external
+    ///   Erlang code. So, since the function doesn't have a fallback pure Gleam
+    ///   implementation, you won't be able to compile it on this target.
+    /// - `uses_javascript_externals: true` the function is using JavaScript
+    ///   external code. This means that you will be able to use it on the
+    ///   JavaScript target with no problems.
+    implementations: Implementations,
+    /// The value's type.
+    #[serde(rename = "type")]
+    type_: TypeInterface,
+}
+
+#[derive(Serialize, Debug)]
+pub struct DeprecationInterface {
+    /// The reason for the deprecation.
+    message: EcoString,
+}
+
+impl DeprecationInterface {
+    fn from_deprecation(deprecation: &Deprecation) -> Option<DeprecationInterface> {
+        match deprecation {
+            Deprecation::NotDeprecated => None,
+            Deprecation::Deprecated { message } => Some(DeprecationInterface {
+                message: message.clone(),
+            }),
+        }
+    }
+}
+
+#[derive(Serialize, Debug)]
+#[serde(tag = "kind")]
+pub enum TypeInterface {
+    /// A tuple type like `#(Int, Float)`.
+    Tuple {
+        /// The types composing the tuple.
+        elements: Vec<TypeInterface>,
+    },
+    /// A function type like `fn(Int, String) -> String`.
+    Fn {
+        parameters: Vec<ParameterInterface>,
+        #[serde(rename = "return")]
+        return_: Box<TypeInterface>,
```

Comment:
Labels are not part of the type system so this information cannot be part of the type AST.

The only places one can have a label is on a module function or on a record constructor so I think we'll need to have fields on those that hold this information.